### PR TITLE
expr/builtin_fncall: implement math.Sin

### PR DESCRIPTION
**Pull request by liufuyang**
## What have you changed? (mandatory)

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
- add math.Sin, as suggested [here](https://pingcap.github.io/blog/adding-built-in-functions-to-tikv/)

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Refer to a related PR or issue link (optional)
https://github.com/tikv/tikv/issues/3275

The related source in tidb: https://github.com/pingcap/tidb/blob/55d8eff75859814114f9c6aa3efe2e3f69f7981f/expression/builtin_math.go#L1676

### CODE REVIEW

**Comment by liufuyang** [comment]
I basically did some copy paste from here>https://github.com/tikv/tikv/pull/3379
If this looks good, I can try add cos in this PR as well

**Comment by liufuyang** [comment]
so are we going to merge this soon or? I can try make a cos afterwards. Or do you expect me to add cos in this PR as well?

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 1]
```diff
@@ -330,4 +337,34 @@ mod test {
...
+            ),
+            (
+                ScalarFuncSig::Sin,
+                Datum::F64(f64::consts::PI / 2.0_f64),
+                Datum::F64(1.0_f64),
+            ),
+            (
+                ScalarFuncSig::Sin,
+                Datum::F64(f64::consts::PI),
+                Datum::F64(0.00000000000000012246467991473532_f64),
```
Comment (vkorenev) [inline src/coprocessor/dag/expr/builtin_math.rs]: **sin(π)** should be equal to 0. While it is clear why the actual result is not precisely equal to 0, I'm afraid that hard-coding it as this particular value may break the build on other architectures.

Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_math.rs]: @zz-jason any suggestion?

**Comment by AndreMouche** [comment]
Maybe you can add other functions in another PR?@liufuyang

File: src/coprocessor/dag/expr/builtin_math.rs (hunk 1 above)
Comment (zz-jason) [reply src/coprocessor/dag/expr/builtin_math.rs]: How about replace the result to `sin(pi())`.

Another thing, in MySQL, `sin(pi())` returns a float-point value nearly to zero, not exactly zero:
```sql

MySQL(localhost:3306) > select sin(pi());
+------------------------+
| sin(pi())              |
+------------------------+
| 1.2246467991473532e-16 |
+------------------------+
1 row in set (0.00 sec)
```

And TiDB behaves the same with MySQL and returns almost the same small value:
```sql
TiDB(localhost:4000) > select sin(pi());
+------------------------+
| sin(pi())              |
+------------------------+
| 1.2246467991473515e-16 |
+------------------------+
1 row in set (0.00 sec)
```

Comment (liufuyang) [reply src/coprocessor/dag/expr/builtin_math.rs]: @vkorenev I assume sin(pi) = 0 is in theory, in computer float calculation it cannot be precise. And thats' also why you normally don't do float comparison with ==.

Should I switch to `sin(pi())`? however I feel if we are going to switch to this, then if rust's implementation of `f64::sin()` changes, then this test will still pass, but we wouldn't know by then the tikv will give a different results than Mysql. (Though this is very very unlikely to happen)

Please let me know what to do now.

Comment (vkorenev) [reply src/coprocessor/dag/expr/builtin_math.rs]: I think the reason why all the programs above return the same result is that **sin** is computed by the same CPU instruction on the same processor family. I think the result may be slightly different on some other processor/arch (AMD?, ARM?) or even on another Intel mircoarchitecture.
Anyway, I'd rather compare the computed value with the mathematically correct result +- some epsilon.

**Comment by liufuyang** [comment]
@vkorenev @zz-jason I updated the test, please take a look again. Let me know what to change.

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 2]
```diff
@@ -343,4 +351,22 @@ mod test {
...
+            (ScalarFuncSig::Sin, Datum::F64(PI / 2.0_f64), 1.0_f64),
+            (ScalarFuncSig::Sin, Datum::F64(PI), 0.0_f64),
+        ];
+        let mut ctx = EvalContext::default();
+        for (sig, arg, exp) in tests {
+            let arg = datum_expr(arg);
+            let f = scalar_func_expr(sig, &[arg]);
+            let op = Expression::build(&mut ctx, f).unwrap();
+            let got = op.eval(&mut ctx, &[]).unwrap();
+            assert!((got.f64() - exp).abs() < 1e-10_f64);
```
[the remark this answers is not in the document]
Comment (vkorenev) [reply src/coprocessor/dag/expr/builtin_math.rs]: Maybe it is better to use either [float_cmp](https://docs.rs/float-cmp/) or [approx](https://docs.rs/approx/) crate for comparing floating point numbers in tests.

[the remark this answers is not in the document]
Comment (liufuyang) [reply src/coprocessor/dag/expr/builtin_math.rs]: @Hoverbear I switched it. @vkorenev Are those crate already imported? I can try use them if people think it is good idea.

File: src/coprocessor/dag/expr/builtin_math.rs [hunk 3]
```diff
@@ -343,4 +351,22 @@ mod test {
...
         }
     }
+
+    #[test]
+    fn test_sin() {
+        let tests = vec![
+            (ScalarFuncSig::Sin, Datum::F64(0.0_f64), 0.0_f64),
+            (ScalarFuncSig::Sin, Datum::F64(PI / 4.0_f64), FRAC_1_SQRT_2),
+            (ScalarFuncSig::Sin, Datum::F64(PI / 2.0_f64), 1.0_f64),
+            (ScalarFuncSig::Sin, Datum::F64(PI), 0.0_f64),
```
Comment (hicqu) [inline src/coprocessor/dag/expr/builtin_math.rs]: How about `f64::NAN` and `f64::INFINITY`?

Comment (liufuyang) [reply src/coprocessor/dag/expr/builtin_math.rs]: You mean test them as input, right?

Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_math.rs]: yes, I think we need them.

**Comment by hicqu** [comment]
LGTM.

**Comment by huachaohuang** [comment]
@solotzg PTAL

**Pull request merged**